cupy/cuda/memory.pyx

```diff
@@ -1434,3 +1434,54 @@ cdef class CFunctionAllocator:
         mem = CFunctionAllocatorMemory(size, self._param, self._malloc_func,
                                        self._free_func, device.get_device_id())
         return MemoryPointer(mem, 0)
+
+
+@cython.no_gc
+cdef class PythonFunctionAllocatorMemory(BaseMemory):
+
+    def __init__(self, size_t size, malloc_func, free_func,
+                 int device_id):
+        self._free_func = free_func
+        self.device_id = device_id
+        self.size = size
+        self.ptr = 0
+        if size > 0:
+            self.ptr = malloc_func(size, device_id)
+
+    def __dealloc__(self):
+        if self.ptr:
+            self._free_func(self.ptr, self.device_id)
+
+
+cdef class PythonFunctionAllocator:
+
+    """Allocator with python functions to perform memory allocation.
+
+    This allocator keeps functions corresponding to *malloc* and *free*,
+    delegating the actual allocation to external sources while only
+    handling the timing of the resource allocation and deallocation.
+
+    *malloc* should follow the signature ``malloc(int, int) -> int``
+    returning the pointer to the allocated memory given the *param* object,
+    the number of bytes to allocate and the device id on which the
+    allocation should take place.
+
+    Similarly, *free* should follow the signature
+    ``free(int, int)`` with no return, taking the pointer to the
+    allocated memory and the device id on which the memory was allocated.
+
```

Comment:
I wonder if this could make it clearer?
```suggestion

    If the external memory management supports asynchronous operations,
    the current CuPy stream can be retrieved inside ``malloc_func`` and
    ``free_func`` by calling :func:`cupy.cuda.get_current_stream()`.

```

Reply:
Thanks!!